# Vectors

Our programs have all been 2 dimensional, so we need to work with both x and y
values, and we've been working with the seperately. Now we can start working
with vectors, which are objects that combine the x and y values. In code,
vectors are usually typle types and look like this: 

```python
(3,5)
```

The first number, `3` is the x value, and the second, `5`, is the y value. 

There are two main ways of thinking about vectors. 

* One is as a position, a point in space which has a X and Y position. 
* The other is as a direction and a magnitude, which is a line with a length and
a direction.

These two ways are actually the same. Any point in space, (x,y) is relative to
the origin, which is the point at (0,0), so for the first way of thinking about
a vector, a point in space, (x, y) we also imiagine that there is a line from
the origin (0,0) to the point (x,y), and that line will have a length and a
direction, which is our second way of thinking about vectors. 

There are a lot of videos and web pages to help you understand vectors in more
depth, [here is a good one](https://youtu.be/VqrYlDcZQ54?feature=shared).

Here is what the vector (x=8, y=8) looks like when we show it on the X/Y coordinate plane, 
starting from the origin. 

<p align="center">
    <img src="images/v88.png" alt="Vector">
</p>


<div style="float: right; width: 250px; margin: 10px 0 10px 25px; padding: 15px; border: 1px solid rgba(128, 128, 128, 0.3); border-radius: 5px; background-color: rgba(128, 128, 128, 0.1); font-style: italic; font-size: 0.85em; opacity: 0.9;">

We are using 2 numbers because our games are 2 dimensional, but vectors can have
more. If we were creating 3D games, our vectors would have 3 numbers, and
mathematicians and physicists often use vectors that have even more numbers.

</div> 

This vector represents the point (8,8) on a grid, and if we draw a line from the
origin (0,0) to the point (8,8), we get a line that is 11.3 units long, at an
angle of 45 degrees. So, it can both be a point and a line, as long as we can
assume a starting point for the line. 

Since we have a new kind of mathematical object, vectors, that have two numbers
in them, like (8,8), we are going to need a new work to refer to regular
numbers, like just plain old '8': these regular numbers are called "scalars".

Here is a really simple Vector class that we can use to represent vectors:



In [10]:
# Run me!


import math

class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    @property
    def magnitude(self): # magnitude is the length of the vector
        # Magnitude of the vector: sqrt(x^2 + y^2)
        return math.sqrt(self.x**2 + self.y**2)

    @property
    def direction(self):
        # Direction (angle) in radians: atan2(y, x)
        return math.atan2(self.y, self.x) * 180 / math.pi

    def __repr__(self):
        return f"Vector({self.x}, {self.y})"

# Example usage:

v = Vector(4, 4)

print(f"Vector: {v}")
print(f"Magnitude: {v.magnitude}")
print(f"Direction: {v.direction}°")


Vector: Vector(4, 4)
Magnitude: 5.656854249492381
Direction: 45.0°




As you can see, your create the vector by passing in the x and y values, and
then you can calculate the length and angle using trigonometry.


There are two very important things that you can do with a vector: we can
multiply a vector by a scalar, and we can add two vectors together. 

* vector * scalar: Multiply the elements of the vector by the scalar
* vector + vector: Add the elements in each position to create a new vector

Let's suppose that we have the vectors (2,3) and (4,5) and the scalar 5.

* (2,3) * 5 = ( 2*5, 3*5) = (10,15)
* (2,3) + (4,5) = ( 2+4, 3+5 ) 

Multiplying a vector by a scalar is called "scaling" the vector; it makes the
vector longer or shorter, but does not change its direction. Adding two vectors
combines the vectors and produces a new vectors that can have a different
magnitude and direction. 



## Assignment

In the code cell below, update our Vector class with implmentations of the special methods for multiplication ( `__mul__`) and 
addition ( `__add__`)

You can check the type of an argument with code like this: 

```python
if isinstance(scalar, (int, float)):
    print('Yes, it is a scalar')
else:
    print('No not a scalar')
```


If the user has provided the wrong types for the arguments, you can signal an error with a line of code like this: 

```python
 raise TypeError("Can only multiply Vector by a scalar (int or float)")
 ```

Both of your implementations must return a new Vector, so they should have code like this:

```python 
return Vector( x_value, y_value)
```

In [12]:
import math

class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    @property
    def magnitude(self): # magnitude is the length of the vector
        # Magnitude of the vector: sqrt(x^2 + y^2)
        return math.sqrt(self.x**2 + self.y**2)

    @property
    def direction(self):
        # Direction (angle) in radians: atan2(y, x)
        return math.atan2(self.y, self.x) * 180 / math.pi

    def __repr__(self):
        return f"Vector({self.x}, {self.y})"
    

    def __mul__(self, scalar):
        """Multiply the vector by a scalar value."""

        pass

    def __add__(self, other):
        pass

# Example usage:
v = Vector(4, 4)

print(f"Vector: {v}")
print(f"Magnitude: {v.magnitude}")
print(f"Direction: {v.direction}°")

# Add more example usage to test the __mul__ and __add__ methods


Vector: Vector(4, 4)
Magnitude: 5.656854249492381
Direction: 45.0°


## PyGame Vector2

Now that you've done all that work creating your own vector class, we aren't going to use it anymore. Sorry, sometimes you write code you don't use, but we wanted you to really understand vectors. 

Instead, we are going to use the Vector class that is built into Pygame: 

```python
from pygame.math import Vector2

```

Suppose that we have a vector called `p1`, which is the position of the player, at
x=100 and y=100. We can write this as:



In [3]:

from pygame.math import Vector2 # Using PyGame's Vector, which has a lot more features

p1 = Vector2(100, 100)

p1 

pygame 2.6.1 (SDL 2.28.4, Python 3.13.3)
Hello from the pygame community. https://www.pygame.org/contribute.html


<Vector2(100, 100)>

When you add vectors, you really just add the x and y values together. 

In [4]:
from  pygame import Vector2

v1 = Vector2(10, 10 )
v2 = Vector2(20, 20 )


print(v1 + v2) # Same as Vector(10+20, 10+20) = Vector(30, 30)

[30, 30]



Now, suppose that we have a vector called `move`, which is the amount we want to
move the player. We can move the player just by adding the move vector to the
player vector. We can write this as:



In [5]:

p1 = Vector2(100, 100)
move = Vector2(10, 0) # Move 10 pixels to the right, none up/down

p1 += move  # same as: p1 = p1 + move
p1

<Vector2(110, 100)>


Now, `p1` is at x=110 and y=100. To get that, we added the x values of the two
vectors together, and the y values of the two vectors together. So the new x
value is 100 + 10 = 110, and the new y value is 100 + 0 = 100.



For a more detaied example. run the program `01a_vector_example.py.` This program has a custom version
of the vector class that displays on a larger grid. Here is what the program's output looks like:

![Vector Example](images/vector_grid.png)


Here are the vectors that are displayed in the program:

```python 
v0 = Vector20(0,0)
v1 = Vector20(8, 8)  
v2 = Vector20(3, -12)  
v3 = Vector20(-4, -2)  
v4 = Vector20(-12, 0) 
v5 = Vector20(0, 12)
```

Compare the code to the output to see how the vectors are displayed on the grid. Notice that we can move from one place
to another on the grid by adding the vectors together. For example, to move from v0 to v1, we can add v0 to v1. 



## Rotations

Since vectors have a direction, it would make sense that we'd want to change the
direction. This operation is called a 'rotation'.

Suppose that we want to move the player 100 pixels in the direction of 45
degrees. The easiest way to write this is to create the vector a 100 units long
in the x direction, then rotate it to 45 degrees. 


In [6]:
p1 = Vector2(100, 100) # Player's current position
move = Vector2(100, 0) # 100 unit long vector
move.rotate_ip(45) # rotate the vector "in-place"

p1 += move

p1

<Vector2(170.711, 170.711)>

Notice the `_ip` suffix after `rotate`; this stands for "in place" and means we are going to change the vector. Without the `_ip` suffix, the `.rotate()` will return a new vector, so we'd have to write it like this: 

```python
move = move.rotate()
```

Rotating a vector changes both the x and y values. The x value is changed by 
`cos(angle) * x`, and the y value is changed by `sin(angle) * y`. 


In [7]:
v1 = Vector2(10,10)
move = Vector2(10, 0).rotate(45)

print(v1)
print(move)

print(v1 + move)


[10, 10]
[7.07107, 7.07107]
[17.0711, 17.0711]


## Assignment 1

Draw a square with Vectors. 

For this assignment we will be using our own vector class, `Vector20` which is
based on `Vector2`, but has extensions for plotting a grid with numbers and
labels. Use it just like you would a `Vector2`

1. Open and read `Vector Example` and run the program.
2. Create vector v1 with x = 0 and y = 1 
3. Scale the vector by 10 with multiplication, then rotate it by 90 degrees so
   it points to the right.
4. Draw the vector from the position (5,5) on the grid.
5. Create a new vector v2 by rotating v1 by 90 dregrees, and draw it from the
   ending position of v1 ( which is the return value from the draw_v20()
   function )
6. Continue rotating and drawing until you have drawn a square. 